File: shared/ui/ipc/host.protocol.notifications.ts
import { NotificationType } from "vscode-jsonrpc";
import { Range } from "vscode-languageserver-types";
import {
	EditorMetrics,
	EditorSelection,
	EditorSidebarLocation,
	IpcRoutes,
	WebviewConfigs
} from "./webview.protocol.common";

/* The following events are expected to be provided from the extension */

export interface ActiveEditorInfo {
	fileName: string;
	languageId?: string;
	uri: string;
	metrics?: EditorMetrics;
	selections: EditorSelection[];
	visibleRanges: Range[];
	lineCount?: number;
}

export interface HostDidChangeActiveEditorNotification {
	editor?: ActiveEditorInfo;
}
export const HostDidChangeActiveEditorNotificationType = new NotificationType<
	HostDidChangeActiveEditorNotification,
	void
>(`${IpcRoutes.Webview}/editor/didChangeActive`);

export interface HostDidChangeLayoutNotification {
	sidebar?: EditorSidebarLocation;
}
export const HostDidChangeLayoutNotificationType = new NotificationType<
	HostDidChangeLayoutNotification,
	void
>(`${IpcRoutes.Webview}/editor/didChangeLayout`);

export interface HostDidChangeVisibleEditorsNotification {
	count: number;
}
export const HostDidChangeVisibleEditorsNotificationType = new NotificationType<
	HostDidChangeVisibleEditorsNotification,
	void
>(`${IpcRoutes.Webview}/editors/didChangeVisible`);

export type HostDidChangeConfigNotification = Partial<WebviewConfigs>;
export const HostDidChangeConfigNotificationType = new NotificationType<
	HostDidChangeConfigNotification,
	void
>(`${IpcRoutes.Webview}/config/didChange`);

export interface HostDidChangeEditorSelectionNotification {
	uri: string;
	selections: EditorSelection[];
	visibleRanges: Range[];
	lineCount?: number;
}
export const HostDidChangeEditorSelectionNotificationType = new NotificationType<
	HostDidChangeEditorSelectionNotification,
	void
>(`${IpcRoutes.Webview}/editor/didChangeSelection`);

export interface HostDidChangeEditorVisibleRangesNotification {
	uri: string;
	selections: EditorSelection[];
	visibleRanges: Range[];
	lineCount?: number;
}
export const HostDidChangeEditorVisibleRangesNotificationType = new NotificationType<
	HostDidChangeEditorVisibleRangesNotification,
	void
>(`${IpcRoutes.Webview}/editor/didChangeVisibleRanges`);

export interface HostDidChangeFocusNotification {
	focused: boolean;
}
export const HostDidChangeFocusNotificationType = new NotificationType<
	HostDidChangeFocusNotification,
	void
>(`${IpcRoutes.Webview}/focus/didChange`);

export interface HostDidLogoutNotification {}
export const HostDidLogoutNotificationType = new NotificationType<HostDidLogoutNotification, void>(
	`${IpcRoutes.Webview}/didLogout`
);

export interface HostDidDidReceiveRequestNotification {
	url?: string;
}

export enum RouteControllerType {
	Codemark = "codemark",
	Review = "review",
	Navigate = "navigate",
	Search = "search",
	PullRequest = "pullRequest",
	StartWork = "startWork"
}
export enum RouteActionType {
	Open = "open"
}
export interface Route {
	controller?: RouteControllerType;
	action?: RouteActionType;
	id?: string;
	query?: any;
}

export const HostDidReceiveRequestNotificationType = new NotificationType<
	HostDidDidReceiveRequestNotification,
	void
>(`${IpcRoutes.Webview}/request/parse`);

File: vs/src/CodeStream.VisualStudio.Core/Models/Host.Protocol.Notifications.cs
using System;
using System.Collections.Generic;

namespace CodeStream.VisualStudio.Core.Models {
	public class HostDidChangeActiveEditorNotificationEditor {
		public HostDidChangeActiveEditorNotificationEditor(string fileName, Uri uri, List<EditorSelection> selections, List<Range> visibleRanges, int? lineCount) {
			FileName = fileName;
			Uri = uri.ToString();
			Selections = selections;
			VisibleRanges = visibleRanges;
			LineCount = lineCount;
		}

		public string FileName { get; }
		public string Uri { get; }
		public List<EditorSelection> Selections { get; }
		public List<Range> VisibleRanges { get; }
		public int? LineCount { get; }

		public string LanguageId { get; set; }
		public EditorMetrics Metrics { get; set; }
	}
	
	public class HostDidChangeActiveEditorNotificationBase {}

	public class HostDidChangeActiveEditorNotification : HostDidChangeActiveEditorNotificationBase {
		public HostDidChangeActiveEditorNotificationEditor Editor { get; set; }
	}

	public class HostDidChangeActiveEditorNotificationType : NotificationType<HostDidChangeActiveEditorNotificationBase> {
		public const string MethodName = "webview/editor/didChangeActive";
		public override string Method => MethodName;
	}

	public class HostDidChangeConfigNotification {
		public bool? ShowHeadshots { get; set; }
		public bool? ShowMarkerGlyphs { get; set; }
		public bool Debug { get; set; }
		public string Email { get; set; }
		public string ServerUrl { get; set; }
	}

	public class HostDidChangeConfigNotificationType : NotificationType<HostDidChangeConfigNotification> {
		public const string MethodName = "webview/config/didChange";
		public override string Method => MethodName;
	}

	public class HostDidChangeEditorSelectionNotification {
		public HostDidChangeEditorSelectionNotification(Uri uri, List<EditorSelection> selections, List<Range> visibleRanges, int? lineCount) {
			Uri = uri.ToString();
			Selections = selections;
			VisibleRanges = visibleRanges;
			LineCount = lineCount;
		}

		public string Uri { get; }
		public List<EditorSelection> Selections { get; }
		public List<Range> VisibleRanges { get; }
		public int? LineCount { get; }
	}

	public class HostDidChangeEditorSelectionNotificationType : NotificationType<HostDidChangeEditorSelectionNotification> {
		public const string MethodName = "webview/editor/didChangeSelection";
		public override string Method => MethodName;
	}

	public class HostDidChangeEditorVisibleRangesNotification {
		public HostDidChangeEditorVisibleRangesNotification(Uri uri, List<EditorSelection> selections, List<Range> visibleRanges, int? lineCount) {
			Uri = uri.ToString();
			Selections = selections;
			VisibleRanges = visibleRanges;
			LineCount = lineCount;
		}

		public string Uri { get; }
		public List<EditorSelection> Selections { get; }
		public List<Range> VisibleRanges { get; }
		public int? LineCount { get; }
	}

	public class HostDidChangeEditorVisibleRangesNotificationType : NotificationType<HostDidChangeEditorVisibleRangesNotification> {
		public const string MethodName = "webview/editor/didChangeVisibleRanges";
		public override string Method => MethodName;
	}

	public class HostDidChangeFocusNotification {
		/// <summary>
		/// Is the IDE focused?
		/// </summary>
		public bool Focused { get; set; }
	}

	public class HostDidChangeFocusNotificationType : NotificationType<HostDidChangeFocusNotification> {
		public const string MethodName = "webview/focus/didChange";
		public override string Method => MethodName;
	}

	public class HostDidLogoutNotification { }
	public class HostDidLogoutNotificationType : NotificationType<HostDidLogoutNotification> {
		public const string MethodName = "webview/didLogout";
		public override string Method => MethodName;
	}

	public class HostDidReceiveRequestNotification {
		public string Url { get; set; }
	}

	public class HostDidReceiveRequestNotificationType : NotificationType<HostDidReceiveRequestNotification> {
		public const string MethodName = "webview/request/parse";
		public override string Method => MethodName;
	}

}
